### PR TITLE
Review Request:  Bavard and Théro

### PR DESCRIPTION
**Pull request by ghost**
**AUTHOR**

Dear @ReScience/editors,

I request a review for the following replication:

### Original article

**Title:**  Adaptive properties of differential learning rates for positive and negative outcomes
**Author(s):**  Cazé, R. D., & van der Meer, M. A. 
**Journal (or Conference):**  Biological cybernetics
**Year:**  2013
**DOI:**  10.1007/s00422-013-0571-5
**PDF:**  [Caze&vanderMeer2013.pdf](https://www.researchgate.net/profile/Romain_Caze/publication/257301717_Adaptive_properties_of_differential_learning_rates_for_positive_and_negative_outcomes/links/0046352c099efc60a0000000/Adaptive-properties-of-differential-learning-rates-for-positive-and-negative-outcomes.pdf)
https://link.springer.com/article/10.1007/s00422-013-0571-5

### Replication

**Author(s)**:   Bavard, Sophie & Théro, Héloïse
**Repository**:  [BAVARD-THERO-2018](https://github.com/heloisethero/ReScience-submission/tree/BAVARD-THERO-2018)
**PDF**:  [Bavard-Thero-2018.pdf](https://github.com/ReScience/ReScience-submission/files/1710600/Bavard-Thero-2018.pdf)
**Keywords**:  Reinforcement learning, Q-learning, adaptive learning rates
**Language**:  Python
**Domain**:  Computational Neuroscience

### Results

* [X] Article has been fully replicated
* [ ] Article has been partially replicated
* [ ] Article has not been replicated

### Potential reviewers 
Benoît Girard
Mehdi Khamassi
Xavier Hinaut

---

**EDITOR**

* [x] Editor acknowledgment (@oliviaguest  `February 20, 2018`)
* [x] Reviewer 1 (@neuronalX `February 21, 2018`)
* [x] Reviewer 2 (@benoit-girard `March 2, 2018`)
* [x] Review 1 decision [accept `May 29, 2018`]
* [x] Review 2 decision [accept `June 11, 2018`]
* [x] Editor decision [accept `June 14, 2018`]

### CODE REVIEW

**Comment by rougier** [comment]
Thanks for your submission. An editor will be assigned soon.

**Comment by rougier** [comment]
@oliviaguest Can you edit this submission ?

**Comment by oliviaguest** [comment]
@rougier I can do it after the 20th of this month, is that too late?

**Comment by rougier** [comment]
@oliviaguest Yes
@heloisethero Can you add a link to your repository (you should be able to edit your first post)

**Comment by rougier** [comment]
@heloisethero Can you also add a link to the PDF? There was a link but it was pointing to the article that seem to be stored on GitHub. Since it is a copyrighted article you cannot add it to GitHub. BEst would be to have an open access version. Else, you can add the official entry from Biol. Cybernetics.

**Comment by ghost** [comment]
@rougier I have now added a link to my repository at the right place, sorry about that. I have also added a link to the first author's ResearchGate article pdf. The figures in it are identical to the ones published in Biol. Cybernetics, although the formatting is different. I have also added to the official entry from Biol. Cybernetics just in case.

**Comment by rougier** [comment]
@heloisethero Perfect. As you may have read in the thread @oliviaguest will handling your submission after February 20th. If this is a problem just tell us, we'll try to find another editor.

**Comment by ghost** [comment]
@rougier and @oliviaguest, there is no problem, Sophie (@sophiebavard) and I accept your decision.

**Comment by rougier** [comment]
@oliviaguest Reminder

**Comment by oliviaguest** [comment]
@benoit-girard @MehdiKhamassi @neuronalX any of you have time to review this?

**Comment by neuronalX** [comment]
@oliviaguest Yes, I can review it.

**Comment by benoit-girard** [comment]
@oliviaguest Technically, I can do it. 
There is however a possible conflict of interest: I am the tutor of H. Théro in her current doctoral school. I do not have any hierarchical relationship with her, I am not supervising her. In this role, I am essentially a possible channel of information when interacting with the doctoral school.
Please tell me if you consider it problematic.
You could also alternatively ask directly to @rcaze (first author of the initial study) to review the submission.

**Comment by MehdiKhamassi** [comment]
@oliviaguest Thank you for the invitation! I am already late with two other reviews. So I cannot currently. Best wishes

**Comment by oliviaguest** [comment]
@MehdiKhamassi @benoit-girard @neuronalX Thank you all for the speedy replies! :smile: 

@rougier What do you think/what are the rules vis-à-vis the potential conflict?

**Comment by rcaze** [comment]
I meant to write to the authors. The python code generating all the figures are on my github repo. They were mislead because they look at Matt Van Der Meer lab repo and could not find the code. I thought the link to the repo was in the original article.

**Comment by rcaze** [comment]
I am going to alert also Matt. https://github.com/rcaze/13_02CaVa

**Comment by ghost** [comment]
@rcaze we are sorry for this mistake. As your article was really influential in our research team, we have tried on several occasions to replicate parts of it. We have been looking for your code, but never found it. Therefore we have recoded everything from scratch.
@rougier @oliviaguest we are sorry for this situation. @sophiebavard and I are ready to withdraw our review request if you think this is the right thing to do.

**Comment by oliviaguest** [comment]
@heloisethero The fact you can replicate the original code using the published papers *without* looking at the original code is extremely important. It means the specification is actually complete given the journal articles and other writing. This is a massive boon to the original authors' work, their framework, etc. For more see, e.g., [Cooper, R. P. & Guest, O. (2014). Implementations are not specifications: specification, replication and experimentation in computational cognitive modeling. Cognitive Systems Research. doi: 10.1016/j.cogsys.2013.05.001.](https://doi.org/10.1016/j.cogsys.2013.05.001)

**Comment by rcaze** [comment]
@heloisethero I agree with @oliviaguest. I am honoured by this attention and think that the independent reproduction gives strength to this result. Notably, the original code lacks a documentation. One member of our lab (@frct) is currently working on this topic and could give a serious review of the work. If the editors wish I can also review the paper even if I think that I am a little biased :).

**Comment by rougier** [comment]
Yes, it is perfectly fine to replicate the code even if the original code is available as long as your replication is not a simple translation from one language to another and is not a copy. So this submission is fine for me.

**Comment by ghost** [comment]
@oliviaguest hello, @sophiebavard and I were wondering if you have chosen the second reviewer?

**Comment by oliviaguest** [comment]
@heloisethero: As soon as @rougier tells me if there's a conflict (see my question above), I'll act accordingly. Understandably, these things can take time.

**Comment by sophiebavard** [comment]
@oliviaguest Thank you for your quick reply :)
I just want to add that @benoit-girard is also my tutor at my doctoral school. He probably didn't mention it because I've only started my PhD 2 month ago and we haven't met yet.

**Comment by rougier** [comment]
@oliviaguest @sophiebavard @heloisethero @benoit-girard Sorry, didn't see the question. @benoit-girard  do you mean "comité de suivi de thèse" when your write tutor? In that case I think it is fine.

**Comment by benoit-girard** [comment]
@sophiebavard : sorry, indeed I did not realize I am your tutor too!
@rougier : not exactly, at ED3C, the tutor is an intermediate between the student and the doctoral school, its main role is to help solving problems if they arise (hopefuly most of the tutored students have no problems). If the student encounters difficulties with the execution of the PhD, and does not feel comfortable alerting directly the doctoral school's head, they can contact the tutor and discuss with him about the problem, the possible solutions, and the opportunity to contact the doctoral school's head.

The tutors are not members of the doctoral school's direction, this is important, because the direction has an interest in having the students publishing as much as possible before the defense. This means that the success or not of the students should not affect the evaluation decisions made by tutors.
But the tutors often (again sorry @sophiebavard) know the student (they usually meet at yearly events organized by the doctoral school, where students present their work with posters or talks), and may thus have developed some sympathy towards them, and would like to see them succeeding. 
I tend to see that as a very very indirect COI, but I am not in position to judge (and in that specific case, we can find a replacement easily if you have a doubt).

**Comment by rougier** [comment]
@benoit-girard Thanks for the explanation.I think it is ok.
@khinsen Do you confirm?

**Comment by khinsen** [comment]
I see no problem with this. In general, conflict of interest is much less of an issue with open reviewing than in traditional publishing. Moreover, we have to measure a particular potential conflict of interest relative to the baseline conflict of interest that is always present among members of relative small communities of practice.

**Comment by rougier** [comment]
@oliviaguest @benoit-girard You're good to go then.

**Comment by benoit-girard** [comment]
OK, I'll start working on it...

**Comment by rougier** [comment]
@oliviaguest @neuronalX @benoit-girard Gentle reminder...

**Comment by benoit-girard** [comment]
I am sorry, I have been very busy, I thought I would be able to do it this week, but won't be able to. 
The next two weeks I will be in holidays and unable to progress on the review. I can resume reviewing at the beginning of May.
@oliviaguest : if you think it is too long a delay, replace me with another reviewer, it is not a problem for me.

**Comment by oliviaguest** [comment]
@benoit-girard: Seems fine to me given we're also waiting for @neuronalX.

**Comment by rougier** [comment]
@neuronalX Any progress on your review?

**Comment by neuronalX** [comment]
@oliviaguest @rougier Sorry for the delay, I have been busy until now. I already told @heloisethero that I was planning to finish by the end of April when we randomly met at a public event in Paris a few weeks ago.
Hopefully I will finish this week.

**Comment by rougier** [comment]
@oliviaguest Any update?

**Comment by oliviaguest** [comment]
@rougier none that I know of. I assume @neuronalX is still working on their review.

**Comment by benoit-girard** [comment]
Sorry for the veeeeeery long delay. I am working on the paper right now.

**Comment by benoit-girard** [comment]
Comments on the paper:
* R Cazé (@rcaze ) has mentionned in the comments above that his original python code is now available on github (https://github.com/rcaze/13_02CaVa), this should be thus added in the last paragraph of the introduction (rather than the link to the empty repository in van der Meer's lab repo).
* the results mention some discrepancies between the original and the reproduced fig. 4, the latter being smoother. I suppose you are refering to Fig 4A? Because fig 4B appears very similar in terms	of jitter. You mention that the number of simulations could be the cause of this discrepancy: could you check with @rcaze how many of these they used,	so as to identify whether it is	the cause? He can be contacted easily.
* Biblio style: the biblio style in ReScience references paper by numbers. Here the authors use a mixed style (name [number]), that I find acceptable. However, some spelling errors have to be corrected: O'doherty -> O'Doherty; O'reilly -> O'Reilly; Cazé and Meer -> Cazé and van der Meer.

Comments on the code:
* the code is simple, clear, easy to read.
* but there is a lot of code replication here (each figure has its own script, thus each script shares a lot of code with the others). It would have been a bit nicer to have a common core code, with parameterized calls for each of the figures. Do you think you could restruturate the code?
* in Figure3.py: why is there a line of "!!!" at line 24 in the comments? Is the comment here incomplete?
* The simulations ran on my computer without problems (Mac OSX 10.10.5, python 2.7.12, numpy 1.11.2, matplotlib 1.5.1),	the results are very similar to those presented in the paper.

**Comment by neuronalX** [comment]
The paper and code are clear and easy to read. I obtain the same figures (except for a detail) when I run the code.
For info, I ran the code on Mac OS X with Python 2.7.13, numpy 1.11.3, matplotlib 2.0.0.

### Random choices
It seems that you generate always the same number of positives (resp. negative) outcomes instead of randomly choosing for each trial (which could lead to different number of positive and negative trial for different simulations). For example, in Figure 1 (the same idea seems to be used for other figures):

  ``` python
  # ConditionIdx refers to the different possible probabilities for positive outcomes
  for conditionIdx in range(len(ProbabilityOfPositiveOutcome)):

    NumberOfPositiveOutcomes = int(ProbabilityOfPositiveOutcome[conditionIdx] * NumberOfTrials)
    NumberOfNegativeOutcomes = NumberOfTrials - NumberOfPositiveOutcomes

    for iterationIdx in range(NumberOfIterations):

      # We create a vector for the outcomes (r = 1 or r = -1) that will be observed on each trial
      Outcomes = np.concatenate((np.ones((NumberOfPositiveOutcomes,), dtype=np.int), - np.ones((NumberOfNegativeOutcomes,), dtype=np.int)))
      random.shuffle(Outcomes)
  ```

Is there a reason to draw a deterministic number of postive trials and shuffle, instead of generating each trial separatly? Or is there something I did not understood?
But elsewere (Figure 2), there is in fact a random choice generated:

  ``` python
  ProbabilityToChooseBestAction = 1/(1+np.exp((Q[0]-Q[1])/beta))
  if random.random() < ProbabilityToChooseBestAction:
    action = 1 # the chosen action is the best one.
    Outcome = OutcomesBestAction[t]
  ```

If there is in fact a reason, that would be great to add some comments about it because it seems unclear to me.

### Figure 3: not the same color for one curve
I obtain another black curve instead of "teal" curve. Do you have an idea where this problem could come from? (matplotlib version?)

figure3


### Figure 4 A: number of simulations
I tried (just once) to change the number of simulations to see if one can obtain a less smoother curve.
I changed the following line from 5000 to 500:

  ``` python
  NumberOfIterations = 5000
  ```

  ``` python
  NumberOfIterations = 500
  ```

figure4_with_500sim-instead-of-5000

With 500 simulations instead of 5000 it seems closer to the figure from the original paper (for part A of the figure only).

### Other general Comments
- could you provide (in the paper, end of Introduction) the Python, numpy, random, matplotlib versions you used for the code?
- could you explain in more details this sentence before giving some examples?
  - "In the reference article, it was sometimes unclear whether the analyses were performed on the probabilities of choice, or rather the proportions of implemented choices."
- "typos":
  - it would seem better to add first sentence of result in the next paragraph
  - could you remove the comma before the semi-colon: "rational and optimistic agents in section 3, i.e.,:"

PS: "funny" that we gave our comments on the same day with @benoit-girard ! And sorry for the delay!

**Comment by ghost** [comment]
Dear @benoit-girard and @neuronalX,

Thank you for your reviews. We have updated our code in our [repository](https://github.com/heloisethero/ReScience-submission/tree/BAVARD-THERO-2018) and here is the new [version of our article](https://github.com/heloisethero/ReScience-submission/blob/BAVARD-THERO-2018/article/Bavard-Thero-2018.pdf).

To answer @benoit-girard’s comments:

- We have changed the last paragraph of the introduction as you recommended.
- About the discrepancies between the original and reproduced fig. 4, we were indeed only referring to figure 4A, as we said in the results: "The only discrepancy we found was in Figure 4 A." We sent an email to @rcaze about the number of iterations for this figure. In the meantime, we saw that with 500 iterations, @neuronalX was able to replicate the jitters of the original figure 4A but not the jitters of the original figure 4B. It is thus possible that in the reference article, a different number of iterations was used to generate the two parts of Figure 4.
- We corrected the spelling mistakes in the references, and also changed the biblio style so that papers are now referenced by numbers only.
- We thank you for your comment about our code structure, and we have now created a script with “core” functions, i.e., functions that are used in almost all the figure script. We still have one script per figure, plus the CoreFunctions script.
- The comment line with “!!!” inside was a mistake, we erased it :)

To answer @neuronalX’s comments:
- About random choices, because we also do behavioural experiments on participants, we are used to pseudo-randomize outcomes. For example, if we want to implement a 75% probability, we ensure that the participant will really see 75% of positive outcomes and 25% of negative outcomes associated with that action. However, we agree that in this case, there is no need to pseudo-randomize. Moreover, pseudo-randomization was not specified in the reference article. Therefore, we have changed our codes to now implement a true randomization for both outcomes and choices.
- About the black color in figure 3, we could find the same problem when we ran the code with Python 2 and not Python 3. The problem seems to come from a difference in division between Python 2 and 3:

`plt.plot(meanPerformanceRational1High, color = ((0/255,128/255,128/255,1)), linewidth=1)
plt.plot(meanPerformanceRational2High, color = ((30/255,50/255,255/255,1)), linewidth=1)
plt.plot(meanPerformanceRational3High, color = ((25/255,25/255,112/255,1)), linewidth=1)`

We have now redefined the colors with Hex triplets, and we have managed to run the code with both Python 2 and 3 without problem. Can you check with the new version of our code?
- We have added at the end of introduction which version of Python, NumPy and Matplotlib we have used. We have also added a more precise description of probabilities and proportions of choice in the methods as you recommended, and we have corrected the typos you noticed.

**Comment by benoit-girard** [comment]
@heloisethero : I am fine with the improvements, thanks a lot. I am checking locally that the simulations are still correct with the new code, I will validate the paper if everything worked fine.

**Comment by benoit-girard** [comment]
fig 1, 3 & 4 are OK
fig 2 has a minor formatting problem:
figure2
on panel C, the high reward P column is partailly cut.

**Comment by ghost** [comment]
Dear @benoit-girard, thank you for your decision. The formatting problem in Figure 2 comes from the matplotlib version you are using (1.5.1). We have changed the code in our repository so now it should work on your matplotlib version as well. Can you check again?

**Comment by benoit-girard** [comment]
There's a problem at line 96:
`plt.bar([3,8], switchPessimistic, width = 1, align = 'center',"" edgecolor = 'k', color='r')`
the "" before edgecolor shouldn't be there, I think...

**Comment by benoit-girard** [comment]
Fig 2 is perfect as soon as these "" are removed...

**Comment by ghost** [comment]
@benoit-girard you are right, thank you. Sorry for that, I have corrected it :)

**Comment by benoit-girard** [comment]
I am now satisfied with the paper.
@oliviaguest : my review is over, and I recommend acceptation.

**Comment by ghost** [comment]
@neuronalX Did you have any chance to check our simulations? :)

**Comment by rougier** [comment]
@oliviaguest @neuronalX Gentle reminder.

**Comment by neuronalX** [comment]
No, sorry, didn't find the time so far to have a complete look at all the comments since my review.

**Comment by rougier** [comment]
@neuronalX Can you give back your comments by the end of the week please.

**Comment by neuronalX** [comment]
Thank you for the changes you made.

> We have now redefined the colors with Hex triplets, and we have managed to run the code with both Python 2 and 3 without problem. Can you check with the new version of our code?

Yes, it works fine for me now.

Concerning the Figure 4 A, maybe you could just add a sentence saying that some similar figure could be obtained by dividing the number of simulations by 10 for instance (so this info does not stay in the review only).

@oliviaguest @rougier I recommend acceptation.

**Comment by oliviaguest** [comment]
Excellent!

@heloisethero do you what to address the small comment above on your figure 4 A?

**Comment by ghost** [comment]
@neuronalX thank you for your comment. We added this sentence at the end of the results section : 

> However, we obtained similar noise when dividing the number of simulations by ten, suggesting the original authors used around 500 iterations for this part of the figure.

@oliviaguest let us know if we need to do anything else?

**Comment by oliviaguest** [comment]
@heloisethero I cannot see anything else that needs doing from your end — so I'll reply on this thread when I am done publishing this, assigning a DOI, etc.

**Comment by oliviaguest** [comment]
This submission has been accepted for publication! It should appear soon on the journal website ([here](http://rescience.github.io/read/)) with DOI: [DOI](https://doi.org/10.5281/zenodo.1289889)
